src/util/debug_backtrace.c

```diff
@@ -0,0 +1,289 @@
+        }
+        _backtrace_printf("   *  ");
+    }
+}
+
+
+
+
+/* ********** Helpers ********** */
+
+
+static inline void _debug_vprintf(const char *format, va_list ap)
+{
+    vfprintf(_sss_debug_file ? _sss_debug_file : stderr, format, ap);
+}
+
+
+static inline void _debug_fwrite(const char *begin, const char *end)
+{
+    if (end <= begin) {
+        return;
+    }
+    size_t size = (end - begin);
+    fwrite_unlocked(begin, size, 1, _sss_debug_file ? _sss_debug_file : stderr);
+}
+
+
+static inline void _debug_fflush(void)
+{
+    fflush(_sss_debug_file ? _sss_debug_file : stderr);
+}
+
+
+ /* does 'level' trigger backtrace dump? */
+static inline bool _is_trigger_level(int level)
+{
+    return ((level <= SSSDBG_OP_FAILURE) &&
+            (level <= debug_level));
+}
+
+
+/* checks if global 'debug_level' has all levels up to 9 enabled */
+static inline bool _all_levels_enabled(void)
+{
+    static const unsigned all_levels =
+      SSSDBG_FATAL_FAILURE|SSSDBG_CRIT_FAILURE|SSSDBG_OP_FAILURE|
+      SSSDBG_MINOR_FAILURE|SSSDBG_CONF_SETTINGS|SSSDBG_FUNC_DATA|
+      SSSDBG_TRACE_FUNC|SSSDBG_TRACE_LIBS|SSSDBG_TRACE_INTERNAL|
+      SSSDBG_TRACE_ALL|SSSDBG_BE_FO;
+
+    unsigned level = debug_level & ~SSSDBG_TRACE_LDB;
+
+    return ((level ^ all_levels) == 0);
+}
+
+
+/* should message of this 'level' go to backtrace? */
+static inline bool _backtrace_is_enabled(int level)
+{
+    /* Store message in backtrace buffer if: */
+    return (_bt.initialized        && /* backtrace is initialized */
+            _bt.enabled            && /* backtrace is enabled */
+            sss_logger != STDERR_LOGGER &&
+            !_all_levels_enabled() && /* generic log doesn't cover everything */
+            level <= SSS_DEBUG_BACKTRACE_LEVEL); /* skip SSSDBG_TRACE_LDB */
+}
+
+
+ /* prints to buffer */
+static void _backtrace_vprintf(const char *format, va_list ap)
+{
+    int buff_tail_size = _bt.size - (_bt.tail - _bt.buffer);
+    int written;
+
+    /* make sure there is at least 1kb available to avoid truncation;
+     * putting a sane limit on the size of single message (1kb in a worst case)
+     * makes logic simpler and avoids performance hit
+     */
+    if (buff_tail_size < 1024) {
+        /* let's wrap */
+        _bt.end = _bt.tail;
+        _bt.tail = _bt.buffer;
+        buff_tail_size = _bt.size;
+    }
+
+    written = vsnprintf(_bt.tail, buff_tail_size, format, ap);
+    if (written >= buff_tail_size) {
+        /* message is > 1kb, just discard */
+        return;
+    }
+
+    _bt.tail += written;
+    if (_bt.tail > _bt.end) {
+        _bt.end = _bt.tail;
+    }
+}
+
+
+static void _backtrace_printf(const char *format, ...)
+{
+    va_list ap;
+    va_start(ap, format);
+    _backtrace_vprintf(format, ap);
+    va_end(ap);
+}
+
+
+static bool _bt_empty(const char *begin, const char *end)
+{
+    int counter = 0;
+
+    while (begin < end) {
+        if (*begin == '\n') {
+            counter++;
+            if (counter == 2) {
```

Comment:
Shall we require greater amount of lines in a buffer to produce a backtrace? For exampe, 5?

Reply:
I think 2 is OK.